### PR TITLE
spirv-fuzz: Add toggle access chain instruction transformation

### PR DESCRIPTION
**Pull request by andreperezmaselco**
In this PR, the classes that represent the toggle access chain instruction transformation and fuzzer pass were implemented. This transformation toggles the instructions `OpAccessChain` and `OpInBoundsAccessChain` between them.

### CODE REVIEW

**Comment by andreperezmaselco** [comment]
Hi, Ally.

For sure.
I'll work on the OpPhi task in the next few days.

Best regards,
André

Em sex., 6 de mar. de 2020 às 08:00, Alastair Donaldson <
notifications@github.com> escreveu:

File: source/fuzz/transformation_toggle_access_chain_instruction.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        const protobufs::InstructionDescriptor& instruction_descriptor) {
+  *message_.mutable_instruction_descriptor() = instruction_descriptor;
+}
+
+bool TransformationToggleAccessChainInstruction::IsApplicable(
+    opt::IRContext* context, const spvtools::fuzz::FactManager& /*unused*/
+    ) const {
+  auto instruction =
+      FindInstruction(message_.instruction_descriptor(), context);
+  if (instruction == nullptr) return false;
```
[the remark this answers is not in the document]
Comment (andreperezmaselco) [reply source/fuzz/transformation_toggle_access_chain_instruction.cpp]: Done.

File: source/fuzz/transformation_toggle_access_chain_instruction.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      message_.instruction_descriptor().target_instruction_opcode());
+
+  assert(instruction->opcode() == opcode &&
+         "The located instruction must have the same opcode as in the "
+         "descriptor.");
+
+  if (opcode == SpvOpAccessChain || opcode == SpvOpInBoundsAccessChain)
+    return true;
+  else
+    return false;
```
[the remark this answers is not in the document]
Comment (andreperezmaselco) [reply source/fuzz/transformation_toggle_access_chain_instruction.cpp]: Done.

File: source/fuzz/transformation_toggle_access_chain_instruction.cpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    opt::IRContext* context, spvtools::fuzz::FactManager* /*unused*/
+    ) const {
+  auto instruction =
+      FindInstruction(message_.instruction_descriptor(), context);
+  SpvOp opcode = instruction->opcode();
+
+  if (opcode == SpvOpAccessChain)
+    instruction->SetOpcode(SpvOpInBoundsAccessChain);
+  else if (opcode == SpvOpInBoundsAccessChain)
+    instruction->SetOpcode(SpvOpAccessChain);
```
[the remark this answers is not in the document]
Comment (andreperezmaselco) [reply source/fuzz/transformation_toggle_access_chain_instruction.cpp]: Done.

File: source/fuzz/transformation_toggle_access_chain_instruction.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+void TransformationToggleAccessChainInstruction::Apply(
+    opt::IRContext* context, spvtools::fuzz::FactManager* /*unused*/
+    ) const {
+  auto instruction =
+      FindInstruction(message_.instruction_descriptor(), context);
+  SpvOp opcode = instruction->opcode();
+
+  if (opcode == SpvOpAccessChain)
```
[the remark this answers is not in the document]
Comment (andreperezmaselco) [reply source/fuzz/transformation_toggle_access_chain_instruction.cpp]: Done.

File: source/fuzz/transformation_toggle_access_chain_instruction.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+class TransformationToggleAccessChainInstruction : public Transformation {
+ public:
+  explicit TransformationToggleAccessChainInstruction(
+      const protobufs::TransformationToggleAccessChainInstruction& message);
+
+  TransformationToggleAccessChainInstruction(
+      const protobufs::InstructionDescriptor& instruction_descriptor);
+
+  // - |message_.instruction_descriptor| must identify an existing
+  // access chain instruction
```
[the remark this answers is not in the document]
Comment (andreperezmaselco) [reply source/fuzz/transformation_toggle_access_chain_instruction.h]: Done.

File: source/fuzz/fuzzer_pass_toggle_access_chain_instruction.cpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+void FuzzerPassToggleAccessChainInstruction::Apply() {
+  auto context = GetIRContext();
+  // Iterates over the module's instructions and checks whether it is
+  // OpAccessChain or OpInBoundsAccessChain. In this case, the transformation is
+  // probabilistically applied.
+  context->module()->ForEachInst([this,
+                                  context](opt::Instruction* instruction) {
+    SpvOp opcode = instruction->opcode();
+    if ((opcode == SpvOpAccessChain || opcode == SpvOpInBoundsAccessChain) &&
+        GetFuzzerContext()->ChooseEven()) {
```
[the remark this answers is not in the document]
Comment (andreperezmaselco) [reply source/fuzz/fuzzer_pass_toggle_access_chain_instruction.cpp]: Done.

**Comment by afd** [comment]
Looks like the CI failed due to a test timeout. @andreperezmaselco Can you pull from master and rebase (necessary anyway due to another transformation having been added), and then I’ll kick off the CI again?

**Pull request merged**